raster-cli/public/PbmImage.hpp

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "Image.hpp"
+
+#include <boost/optional.hpp>
+
+#include <fstream>
+
+class Operation;
+
+class PbmImage : public Image {
+public:
+  PbmImage(const std::string& file_path);
+
+  void apply(const Operation& operation) override;
+
+  virtual std::ostream& output(std::ostream& out) const = 0;
+  virtual std::istream& input(std::istream& in) = 0;
+
+public:
+  bool is_loaded() const;
+
+  const std::string& get_format_id() const;
+  const size_t get_width() const;
+  const size_t get_height() const;
+
+protected:
+  void set_format_id(const std::string& format_id);
+  void set_width(size_t width);
+  void set_height(size_t height);
+
+private:
```

Comment:
I'm not sure how optional are these

Reply:
You're right the semantics is wrong here. I used boost::optional for these because it provides a handy way to check if the var is initialized. I needed this because the setters allow the separate loading of the data in the image which means that I have to track them separately to check if they are all loaded. May be the solution will be to have a wrapper structure around this data fields e.g. Metadata. But the Monochrome type doesn't have one of the fields the other two have - the max value for a pixel. I think for now I'll alias boost::optional to something more semantically correct and will continue forward. If time allows we can rethink this part of the code. I think it is a reasonable solution for now.